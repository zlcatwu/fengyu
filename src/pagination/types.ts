export type IPaginationOptions = {
  limit: number;
  page: number;
  total: number;
}

export const paginationProps = {
  limit: {
    type: Number,
    default: 10
  },
  page: {
    type: Number,
    default: 1
  },
  total: {
    type: Number,
    default: 0
  }
}
