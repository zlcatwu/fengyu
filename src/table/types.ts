/* eslint-disable @typescript-eslint/no-explicit-any */

import { VNode } from 'vue'
import type { PropOptions, PropType } from 'vue-types/dist/types'
type Prop<T, D = T> = PropOptions<T, D> | PropType<T>
type PublicRequiredKeys<T> = {
  [K in keyof T]: T[K] extends { required: true } ? K : never
}[keyof T]

type PublicOptionalKeys<T> = Exclude<keyof T, PublicRequiredKeys<T>>
type InferPropType<T> = T extends null
  ? any // null & true would fail to infer
  : T extends { type: null | true }
    ? any // As TS issue https://github.com/Microsoft/TypeScript/issues/14829 // somehow `ObjectConstructor` when inferred from { (): T } becomes `any` // `BooleanConstructor` when inferred from PropConstructor(with PropMethod) becomes `Boolean`
    : T extends ObjectConstructor | { type: ObjectConstructor }
      ? Record<string, any>
      : T extends BooleanConstructor | { type: BooleanConstructor }
        ? boolean
        : T extends Prop<infer V, infer D>
          ? unknown extends V
            ? D
            : V
          : T

// eslint-disable-next-line @typescript-eslint/ban-types
export type IxPublicPropTypes<O> = O extends object
  ? { [K in PublicRequiredKeys<O>]: InferPropType<O[K]> } & { [K in PublicOptionalKeys<O>]?: InferPropType<O[K]> }
  : { [K in string]: any }

export enum SORT_TYPE {
  NONE = 'NONE',
  ASC = 'ASC',
  DESC = 'DESC'
}

export enum SELECT_TYPE {
  NONE = 'NONE',
  MULTI = 'MULTI',
  SINGLE = 'SINGLE'
}

export type ITableData = {
  [key: string]: any;
}

export type Slot = (...args: any[]) => VNode[]

export type Slots = {
  [key: string]: Slot | undefined
};

// type ITableOptions = {
// }

// export type ISelectOptions = {
//   idProperty: string;
//   selected: Array<String | Number>;
//   selectType: SELECT_TYPE;
//   onBeforeSelect?: () => boolean;
//   renderFn?: () => VNode | void;
// }

export type ISortOptions = {
  sortKey: string;
  sortType: SORT_TYPE;
  remote: boolean;
}

export type IColumnOptions = {
  header: string;
  dataIndex: string;
  // visible?: boolean;
  sortable?: boolean;
  sortFn: (ra: ITableData, rb: ITableData) => number;
  slot?: ITableHeaderSlot & ITableContentSlot;
}

// Props 定义在这里
export const tableProps = {
  data: {
    type: Array as PropType<Array<ITableData>>,
    default: () => []
  },
  // options: {
  //   type: Object as PropType<ITableOptions>,
  //   default: () => ({})
  // },
  // selectOptions: {
  //   type: Object as PropType<ISelectOptions>,
  //   default: () => ({})
  // },
  paginationOptions: {
    type: Object as PropType<ITablePaginationOptions>,
    default: () => ({})
  },
  sortOptions: {
    type: Object as PropType<ISortOptions>,
    default: () => ({})
  },
  columns: {
    type: Array as PropType<Array<IColumnOptions>>,
    default: () => []
  }
}

export const tableHeaderProps = {
  columns: {
    type: Array as PropType<Array<IColumnOptions>>,
    default: () => []
  },
  sortOptions: {
    type: Object as PropType<ISortOptions>,
    default: () => ({})
  }
}

export const tableBodyProps = {
  columns: {
    type: Array as PropType<Array<IColumnOptions>>,
    default: () => []
  },
  data: {
    type: Array as PropType<Array<ITableData>>,
    default: () => []
  }
}

export const tableBodyCellProps = {
  column: {
    type: Object as PropType<IColumnOptions>,
    default: () => ({})
  },
  data: {
    type: Object as PropType<ITableData>,
    default: () => ({})
  }
}

export const tableHeaderCellProps = {
  column: {
    type: Object as PropType<IColumnOptions>,
    default: () => ({})
  },
  sortOptions: {
    type: Object as PropType<ISortOptions>,
    default: () => ({})
  }
}

// 表格的分页需要关注 enbale 与 remote，但分页器组件不需要关注
export type ITablePaginationOptions = {
  enable: boolean;
  limit: number;
  page: number;
  total: number;
  remote?: boolean;
}

export type ITableHeaderSlot = (options: {
  value: any,
  data: IColumnOptions
}) => VNode;
export type ITableContentSlot = (options: {
  value: any,
  data: ITableData
}) => VNode;

export type TablePublicProps = IxPublicPropTypes<typeof tableProps>
