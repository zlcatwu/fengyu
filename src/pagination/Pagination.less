.fy-pagination {
  margin-top: 12px;
  text-align: center;

  &__prev,
  &__next {
    color: blue;
    cursor: pointer;

    &.disabled {
      cursor: unset;
      color: grey;
    }
  }

  &__next,
  &__page {
    margin-left: 12px;
  }
}
