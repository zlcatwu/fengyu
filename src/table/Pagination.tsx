/**
 * Created by uedc on 2021/10/11.
 */

import { computed, defineComponent } from '@vue/composition-api'
import { paginationProps } from './types'
import './Table.less';

export default defineComponent({
  name: 'FyTable',
  props: paginationProps,
  setup(props, { emit }) {
    const pageTotal = computed(() => Math.ceil(props.total / props.limit));
    const onTogglePage = (targetPage: number) => {
      if (targetPage < 0 || targetPage > pageTotal.value) {
        return;
      }
      emit('update:page', targetPage);
    };

    return {
      pageTotal,
      onTogglePage
    };
  },
  render() {
    return <div class="fy-table-pagination">
      <span
        onClick={() => this.onTogglePage(this.page - 1)}
        class="fy-table-pagination__prev">
          prev
      </span>
      <span
        onClick={() => this.onTogglePage(this.page + 1)}
        class="fy-table-pagination__next">
          next
      </span>
      <span
        class="fy-table-pagination__page">
          ({this.page} / {this.pageTotal})
      </span>
    </div>
  }
})
