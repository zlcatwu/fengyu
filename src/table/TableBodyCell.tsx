/**
 * Created by uedc on 2021/10/11.
 */

import { computed, defineComponent } from '@vue/composition-api'
import { tableBodyCellProps } from './types';

export default defineComponent({
    name: 'TableBodyCell',
    props: tableBodyCellProps,
    setup(props) {
      const value = computed(() => props.data[props.column.dataIndex]);

      return {
        value
      };
    },
    render() {
        return <td class="fy-table__body-column">
        <span class="fy-table__body-column__value">
          {this.column.slot ? this.column.slot({ value: this.value, data: this.data }) : this.value}
        </span>
      </td>
    }
})

