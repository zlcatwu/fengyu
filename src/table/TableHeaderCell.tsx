/**
 * Created by uedc on 2021/10/11.
 */

import { computed, defineComponent } from '@vue/composition-api'
import { TRACE } from '../utils';
import { tableHeaderCellProps, ISortOptions, SORT_TYPE, IColumnOptions } from './types';

export default defineComponent({
  name: 'TableHeaderCell',
  props: tableHeaderCellProps,
  setup(props, { emit }) {
    const { nextType, sortCls } = useSortStates(props);
    const onSortChange = () => {
      if (!props.column.sortable) {
          return;
      }
      const nextSortOptions: ISortOptions = {
          sortType: nextType.value,
          sortKey: props.column.dataIndex,
          remote: props.sortOptions.remote
      };
      TRACE({
        msg: `onSortChange trigger update:sortOptions ${JSON.stringify(nextSortOptions)}`,
        module: 'TableHeaderCell'
      });
      emit('update:sortOptions', nextSortOptions);
    };
    return {
      onSortChange,
      sortCls
    };
  },
  render() {
    return <th onClick={() => this.onSortChange()} class="fy-table__header-column">
      <span class="fy-table__header-column__value">
        {this.column.slot ? this.column.slot({
          value: this.column.header,
          data: this.column
        }) : this.column.header}
      </span>
      {this.column.sortable && <span class={this.sortCls}></span>}
    </th>
  }
})

function useSortStates(options: {
  column: IColumnOptions,
  sortOptions: ISortOptions
}) {
  const sortClsMap = {
      [SORT_TYPE.NONE]: 'fy-table__header-column_sortable',
      [SORT_TYPE.ASC]: 'fy-table__header-column_sortable_asc',
      [SORT_TYPE.DESC]: 'fy-table__header-column_sortable_desc',
  };
  const nextSortTypeMap = {
      [SORT_TYPE.NONE]: SORT_TYPE.ASC,
      [SORT_TYPE.ASC]: SORT_TYPE.DESC,
      [SORT_TYPE.DESC]: SORT_TYPE.NONE
  };
  const curType = computed(() => options.column.dataIndex === options.sortOptions.sortKey ? options.sortOptions.sortType : SORT_TYPE.NONE);
  const nextType = computed(() => nextSortTypeMap[curType.value]);
  const sortCls = computed(() => sortClsMap[options.column.dataIndex === options.sortOptions.sortKey ? options.sortOptions.sortType : SORT_TYPE.NONE]);
  return {
      sortCls,
      nextType
  };
}
