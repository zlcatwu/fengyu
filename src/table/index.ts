/**
 * Created by uedc on 2021/10/11.
 */

import TestTable from './Table'

export { TestTable }
