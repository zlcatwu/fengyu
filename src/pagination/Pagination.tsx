/**
 * Created by uedc on 2021/10/11.
 */

import { computed, defineComponent } from '@vue/composition-api'
import { IPaginationOptions, paginationProps } from './types'
import './Pagination.less';
import { INFO, TRACE } from '../utils';

export default defineComponent({
  name: 'Pagination',
  props: paginationProps,
  setup(props, { emit }) {
    const pageTotal = computed(() => Math.ceil(props.total / props.limit));
    const { prevCls, nextCls } = useClasses(props, pageTotal.value);
    const onTogglePage = (targetPage: number) => {
      if (targetPage <= 0 || targetPage > pageTotal.value) {
        return;
      }
      TRACE({
        msg: `onTogglePage trigger update:page ${targetPage}`,
        module: 'Pagination'
      });
      emit('update:page', targetPage);
    };

    return {
      pageTotal,
      onTogglePage,
      prevCls,
      nextCls
    };
  },
  render() {
    INFO({
      msg: `render with page: ${this.page}, total: ${this.total}, limit: ${this.limit}, pageTotal: ${this.pageTotal}`,
      module: 'Pagination'
    });
    return <div class="fy-pagination">
      <span
        onClick={() => this.onTogglePage(this.page - 1)}
        class={this.prevCls}>
          prev
      </span>
      <span
        onClick={() => this.onTogglePage(this.page + 1)}
        class={this.nextCls}>
          next
      </span>
      <span
        class="fy-pagination__page">
          ({this.page} / {this.pageTotal})
      </span>
    </div>
  }
})

function useClasses (options: IPaginationOptions, pageTotal: number) {
  const prevCls = computed(() => `fy-pagination__prev ${options.page <= 1 ? 'disabled' : ''}`);
  const nextCls = computed(() => {
    return `fy-pagination__next ${options.page >= pageTotal ? 'disabled' : ''}`;
  });
  return {
    prevCls,
    nextCls
  };
}
