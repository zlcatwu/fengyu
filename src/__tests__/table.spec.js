import { mount } from '@vue/test-utils'
import { FyTable } from '../table'
import TableBodyCell from '../table/TableBodyCell'
import TableHeaderCell from '../table/TableHeaderCell'
import { SORT_TYPE } from '../table/types'
import Pagination from '../pagination/Pagination'

describe('Table', () => {
  const TableMount = options => mount(FyTable, options)

  test('render', () => {
    const wrapper = TableMount()
    expect(wrapper.html()).toMatchSnapshot()
    expect(() => {
      wrapper.vm.$forceUpdate()
      wrapper.vm.$destroy()
    }).not.toThrow()
  })

  test('options with sample columns and data', () => {
    const wrapper = TableMount({
      propsData: {
        columns: [
          { dataIndex: 'name', header: 'Name' },
          { dataIndex: 'account', header: 'Account' }
        ],
        data: [
          { name: 'name-1', account: 'account-1' },
          { name: 'name-2', account: 'account-2' }
        ]
      }
    });
    expect(wrapper.findAllComponents(TableHeaderCell).at(0).html().includes('Name')).toBeTruthy();
    expect(wrapper.findAllComponents(TableHeaderCell).at(1).html().includes('Account')).toBeTruthy();
    expect(wrapper.findAllComponents(TableBodyCell).at(0).html().includes('name-1')).toBeTruthy();
    expect(wrapper.findAllComponents(TableBodyCell).at(1).html().includes('account-1')).toBeTruthy();
    expect(wrapper.findAllComponents(TableBodyCell).at(2).html().includes('name-2')).toBeTruthy();
    expect(wrapper.findAllComponents(TableBodyCell).at(3).html().includes('account-2')).toBeTruthy();
  });

  test('sort: sortable class should exits', () => {
    const wrapper = TableMount({
      propsData: {
        columns: [
          { dataIndex: 'name', header: 'Name', sortable: true }
        ]
      }
    });
    expect(wrapper.findAllComponents(TableHeaderCell).at(0).html().includes('fy-table__header-column_sortable')).toBeTruthy();
  });

  test('sort: built-in sort', async () => {
    const data = [
        { name: 'name-5', rank: 5 },
        { name: 'name-3', rank: 3 },
        { name: 'name-2', rank: 2 },
        { name: 'name-4', rank: 4 },
        { name: 'name-1', rank: 1 }
    ];
    const wrapper = TableMount({
      propsData: {
        columns: [
          { dataIndex: 'name', header: 'Name' },
          { dataIndex: 'rank', header: 'Rank', sortable: true }
        ],
        data: [
          ...data
        ],
        sortOptions: {
          sortKey: '',
          sortType: SORT_TYPE.NONE
        }
      }
    });
    let sortedData = [...data];
    // .sync 写法单纯触发事件不会更新视图，需要手动 setProps 一下
    wrapper.vm.$on('update:sortOptions', (opt) => {
      wrapper.setProps({
        sortOptions: opt
      })
    });
    // 点一次升序
    await wrapper.findAllComponents(TableHeaderCell).at(1).trigger('click');
    sortedData.sort((a, b) => a.rank > b.rank ? 1 : -1);
    for (let i = 0; i < sortedData.length; i++) {
      expect(wrapper.findAllComponents(TableBodyCell).at(i * 2).html().includes(sortedData[i].name)).toBeTruthy();
      expect(wrapper.findAllComponents(TableBodyCell).at(i * 2 + 1).html().includes(sortedData[i].rank)).toBeTruthy();
    }
    // 再点一次降序
    await wrapper.findAllComponents(TableHeaderCell).at(1).trigger('click');
    sortedData.sort((a, b) => a.rank > b.rank ? -1 : 1);
    for (let i = 0; i < sortedData.length; i++) {
      expect(wrapper.findAllComponents(TableBodyCell).at(i * 2).html().includes(sortedData[i].name)).toBeTruthy();
      expect(wrapper.findAllComponents(TableBodyCell).at(i * 2 + 1).html().includes(sortedData[i].rank)).toBeTruthy();
    }
    // 再点一次不排序
    await wrapper.findAllComponents(TableHeaderCell).at(1).trigger('click');
    for (let i = 0; i < data.length; i++) {
      expect(wrapper.findAllComponents(TableBodyCell).at(i * 2).html().includes(data[i].name)).toBeTruthy();
      expect(wrapper.findAllComponents(TableBodyCell).at(i * 2 + 1).html().includes(data[i].rank)).toBeTruthy();
    }

    // 点击 sortable: false 的列，应该无效果
    await wrapper.findAllComponents(TableHeaderCell).at(0).trigger('click');
    for (let i = 0; i < data.length; i++) {
      expect(wrapper.findAllComponents(TableBodyCell).at(i * 2).html().includes(data[i].name)).toBeTruthy();
      expect(wrapper.findAllComponents(TableBodyCell).at(i * 2 + 1).html().includes(data[i].rank)).toBeTruthy();
    }
  });

  test('sort: custom sort', async () => {
    const data = [
      { name: 'name-1', rank: 5 },
      { name: 'name-2', rank: 3 },
      { name: 'name-3', rank: 2 },
      { name: 'name-4', rank: 4 },
      { name: 'name-5', rank: 1 }
    ];
    const sortFn = (a, b) => a.rank < b.rank ? 1 : -1;
    const wrapper = TableMount({
      propsData: {
        columns: [
          { dataIndex: 'name', header: 'Name' },
          { dataIndex: 'rank', header: 'Rank', sortable: true, sortFn }
        ],
        data
      }
    });
    // .sync 写法单纯触发事件不会更新视图，需要手动 setProps 一下
    wrapper.vm.$on('update:sortOptions', (opt) => {
      wrapper.setProps({
        sortOptions: opt
      })
    });
    await wrapper.findAllComponents(TableHeaderCell).at(1).trigger('click');
    const sortedData = [...data];
    sortedData.sort(sortFn);
    for (let i = 0; i < sortedData.length; i++) {
      expect(wrapper.findAllComponents(TableBodyCell).at(i * 2).html().includes(sortedData[i].name)).toBeTruthy();
      expect(wrapper.findAllComponents(TableBodyCell).at(i * 2 + 1).html().includes(sortedData[i].rank)).toBeTruthy();
    }

    await wrapper.findAllComponents(TableHeaderCell).at(1).trigger('click');
    sortedData.sort((a, b) => -sortFn(a, b));
    for (let i = 0; i < sortedData.length; i++) {
      expect(wrapper.findAllComponents(TableBodyCell).at(i * 2).html().includes(sortedData[i].name)).toBeTruthy();
      expect(wrapper.findAllComponents(TableBodyCell).at(i * 2 + 1).html().includes(sortedData[i].rank)).toBeTruthy();
    }

    await wrapper.findAllComponents(TableHeaderCell).at(1).trigger('click');
    for (let i = 0; i < data.length; i++) {
      expect(wrapper.findAllComponents(TableBodyCell).at(i * 2).html().includes(data[i].name)).toBeTruthy();
      expect(wrapper.findAllComponents(TableBodyCell).at(i * 2 + 1).html().includes(data[i].rank)).toBeTruthy();
    }
  })

  test('sort: remote sort', async () => {
    const data = [
      { name: 'name-1', rank: 5 },
      { name: 'name-2', rank: 3 },
      { name: 'name-3', rank: 2 },
      { name: 'name-4', rank: 4 },
      { name: 'name-5', rank: 1 }
    ];
    const wrapper = TableMount({
      propsData: {
        columns: [
          { dataIndex: 'name', header: 'Name' },
          { dataIndex: 'rank', header: 'Rank', sortable: true }
        ],
        data,
        sortOptions: {
          sortKey: '',
          sortType: SORT_TYPE.NONE,
          remote: true
        }
      }
    });
    // .sync 写法单纯触发事件不会更新视图，需要手动 setProps 一下
    wrapper.vm.$on('update:sortOptions', (opt) => {
      wrapper.setProps({
        sortOptions: opt
      })
    });
    // 配了 remote 的点击后不会引起数据更新，单纯外发事件交由外部处理
    await wrapper.findAllComponents(TableHeaderCell).at(1).trigger('click');
    for (let i = 0; i < data.length; i++) {
      expect(wrapper.findAllComponents(TableBodyCell).at(i * 2).html().includes(data[i].name)).toBeTruthy();
      expect(wrapper.findAllComponents(TableBodyCell).at(i * 2 + 1).html().includes(data[i].rank)).toBeTruthy();
    }
  });

  test('slot', () => {
    const wrapper = TableMount({
      propsData: {
        columns: [
          { dataIndex: 'name', header: 'Name' },
          { dataIndex: 'account', header: 'Account' }
        ],
        data: [
          { name: 'name-1', account: 'account-1' },
          { name: 'name-2', account: 'account-2' }
        ],
      },
      slots: {
        header__name: '<div>NAME</div>'
      },
      scopedSlots: {
        body__name: '<div slot-scope="scope">[ {{scope.data.name.toUpperCase()}} ]</div>'
      }
    });
    expect(wrapper.findAllComponents(TableHeaderCell).at(0).html().includes('NAME')).toBeTruthy();
    expect(wrapper.findAllComponents(TableBodyCell).at(0).html().includes('NAME-1')).toBeTruthy();
  });

  test('pagination: limit < data.length', () => {
    const wrapper = TableMount({
      propsData: {
        columns: [
          { dataIndex: 'name', header: 'Name', sortable: true }
        ],
        data: [
          { name: 'xxx' },
          { name: 'yyy' },
          { name: 'zzz' }
        ],
        paginationOptions: {
          enable: true,
          limit: 5,
          page: 1
        }
      }
    });
    expect(wrapper.findComponent(Pagination).exists()).toBeTruthy();
  });

  test('pagination: limit > data.length', async () => {
    const wrapper = TableMount({
      propsData: {
        columns: [
          { dataIndex: 'name', header: 'Name' }
        ],
        data: [
          { name: 'name-1' },
          { name: 'name-2' },
          { name: 'name-3' }
        ],
        paginationOptions: {
          enable: true,
          limit: 2,
          page: 1
        }
      }
    });
    // 共2页，当前第1页
    expect(wrapper.findComponent(Pagination).html().includes('(1 / 2)')).toBeTruthy();
    // .sync 写法单纯触发事件不会更新视图，需要手动 setProps 一下
    wrapper.vm.$on('update:paginationOptions', (opt) => {
      wrapper.setProps({
        paginationOptions: opt
      })
    });

    // 在第一页点上一页，应该无效果
    await wrapper.findComponent(Pagination).find('.fy-pagination__prev').trigger('click');
    expect(wrapper.findAllComponents(TableBodyCell).at(0).html().includes('name-1')).toBeTruthy();

    // 正常跳转下一页
    await wrapper.findComponent(Pagination).find('.fy-pagination__next').trigger('click');
    expect(wrapper.findAllComponents(TableBodyCell).at(0).html().includes('name-3')).toBeTruthy();

    // 在最后一页点下一页，应该无效果
    await wrapper.findComponent(Pagination).find('.fy-pagination__next').trigger('click');
    expect(wrapper.findAllComponents(TableBodyCell).at(0).html().includes('name-3')).toBeTruthy();
  });

  test('pagination: remote pagination', async () => {
    const wrapper = TableMount({
      propsData: {
        columns: [
          { dataIndex: 'name', header: 'Name' }
        ],
        data: [
          { name: 'name-1' },
          { name: 'name-2' }
        ],
        paginationOptions: {
          limit: 2,
          page: 1,
          total: 10,
          remote: true,
          enable: true
        }
      }
    });
    // .sync 写法单纯触发事件不会更新视图，需要手动 setProps 一下
    wrapper.vm.$on('update:paginationOptions', (opt) => {
      wrapper.setProps({
        paginationOptions: opt
      })
    });

    expect(wrapper.findComponent(Pagination).html().includes('1 / 5')).toBeTruthy();

    // 有下一页，但点击只触发事件，数据更新交由外部去做
    await wrapper.findComponent(Pagination).find('.fy-pagination__next').trigger('click');
    expect(wrapper.findAllComponents(TableBodyCell).at(0).html().includes('name-1')).toBeTruthy();
  });

})
