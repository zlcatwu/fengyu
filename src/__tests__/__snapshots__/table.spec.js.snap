// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Table render 1`] = `
"<div class=\\"fy-table\\">
  <table class=\\"fy-table__table\\">
    <thead>
      <tr class=\\"fy-table__header-row\\"></tr>
    </thead>
    <tbody></tbody>
  </table>
</div>"
`;
